### PR TITLE
HTTP Basic Auth

### PR DESCRIPTION
**Pull request by szareiangm**
Hello team,

I noticed that HTTP basic auth is becoming a more popular option for Elasticsearch specially with their free offering of x-pack free license. So, I made this contribution. Hope that it helps

CC: @nafid

### CODE REVIEW

File: http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala [hunk 1]
```diff
@@ -56,8 +64,18 @@ class ElasticsearchSenderHTTP(
...
+
+  private val formedHost = new HttpHost(endpoint, port,if (uri.options.getOrElse("ssl", "false") == "true") "https" else "http")
+  private val restClientBuilder = RestClient.builder(formedHost)
+    .setHttpClientConfigCallback(httpClientConfigCallback.asInstanceOf[RestClientBuilder.HttpClientConfigCallback])
+
+  if (username.orNull != null && password.orNull != null) {
+    val userpass = BaseEncoding.base64().encode(s"${username.get}:${password.get}".getBytes(Charsets.UTF_8))
+    val headers:Array[Header] = Array(new BasicHeader("Authorization", s"Basic $userpass"))
+    restClientBuilder.setDefaultHeaders(headers)
+  }
```
Comment (BenFradet) [inline http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala]: you can do:

```scala
private val client = {
  val httpClientConfigCallback =
     if (awsSigning) new SignedHttpClientConfigCallback(credentialsProvider, region)
     else NoOpHttpClientConfigCallback
  val formedHost = new HttpHost(endpoint, port, if (ssl) "https" else "http")
  val headers = (username, password) match {
    case (Some(u), Some(p)) =>
      val userpass = BaseEncoding.base64().encode(s"${username.get}:${password.get}".getBytes(Charsets.UTF_8))
      Array(new BasicHeader("Authorization", s"Basic $userpass"))
    case _ => Array.empty
  }
  val builder = RestClient.builder(formedHost)
    .setHttpClientConfigCallback(httpClientConfigCallback)
    .setDefaultHeaders(headers)
  HttpClient.fromRestClient(restClientBuilder.build())
}
```

File: http/src/test/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTPSpec.scala [hunk 2]
```diff
@@ -32,7 +32,9 @@ class ElasticsearchSenderHTTPSpec extends Specification {
   node.start()
   val client = node.elastic4sclient()
   val creds = CredentialsLookup.getCredentialsProvider("a", "s")
-  val sender = new ElasticsearchSenderHTTP(node.ip, node.port, creds, "region", false, false, None, 1000L, 1)
+  val username:Option[String] = None: Option[String]
+  val password:Option[String] = None: Option[String]
+  val sender = new ElasticsearchSenderHTTP(node.ip, node.port, username, password, creds, "region", false, false, None, 1000L, 1)
```
Comment (BenFradet) [inline http/src/test/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTPSpec.scala]: I think it's fine to do: ```new ElasticsearchSenderHTTP(node.ip, node.port, None, None, creds, "region", false, false, None, 1000L, 1)``` directly

File: http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala [hunk 3]
```diff
@@ -56,8 +64,18 @@ class ElasticsearchSenderHTTP(
   private val httpClientConfigCallback =
     if (awsSigning) new SignedHttpClientConfigCallback(credentialsProvider, region)
     else NoOpHttpClientConfigCallback
-  private val client = HttpClient(uri,
-    httpClientConfigCallback = httpClientConfigCallback)
+
+  private val formedHost = new HttpHost(endpoint, port,if (uri.options.getOrElse("ssl", "false") == "true") "https" else "http")
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala]: addressed. was in my commits but I missed to push. sorry about it.

File: http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala [hunk 4]
```diff
@@ -56,8 +64,18 @@ class ElasticsearchSenderHTTP(
...
+  private val restClientBuilder = RestClient.builder(formedHost)
+    .setHttpClientConfigCallback(httpClientConfigCallback.asInstanceOf[RestClientBuilder.HttpClientConfigCallback])
```
[the remark this answers is not in the document]
Comment (szareiangm) [reply http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala]: Not really necessary

File: http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala (hunk 1 above)
Comment (szareiangm) [reply http/src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTP.scala]: thanks. I have put it in the code.

File: http/src/test/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTPSpec.scala (hunk 2 above)
Comment (szareiangm) [reply http/src/test/scala/com.snowplowanalytics.snowplow.storage.kinesis.elasticsearch/clients/ElasticsearchSenderHTTPSpec.scala]: sure thing. done

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

📝 Please visit https://github.com/snowplow/snowplow/wiki/CLA to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

**Comment by snowplowcla** [comment]
@szareiangm has signed the Software Grant and Corporate Contributor License Agreement

**Comment by BenFradet** [comment]
Thanks, merged!

**Pull request merged**